chalice/chalicelib/request_handler.py

```diff
@@ -0,0 +1,128 @@
+
+
+class RequestHandler:
+    s3 = boto3.resource("s3")
+
+    @staticmethod
+    def generate_request_id(bundle_uuids: List[str]) -> str:
+        """
+        Generate a request id based on a list of bundle uuids.
+        :param bundle_uuids: A list of bundle uuids.
+        :return: Request id.
+        """
+        bundle_uuids.sort()
+        m = hashlib.sha256()
+
+        for bundle_uuid in bundle_uuids:
+            m.update(bundle_uuid.encode())
+
+        return m.hexdigest()
+
+    @staticmethod
+    def get_request_field(request_id: str, fieldname: str) -> str:
+        """
+        Get a field value from a request status file in s3.
+        :param request_id: Matrices concatenation request id.
+        :param fieldname: Field name in the request body
+        :return: Corresponding field value in the request body.
+        """
+        key = request_id + JSON_SUFFIX
+
+        try:
+            content = s3_blob_store.get(bucket=REQUEST_STATUS_BUCKET_NAME, key=key)
+            value = json.loads(content)[fieldname]
+            return value
+        except Exception as e:
+            raise e
+
+    @staticmethod
+    def get_request_body(request_id: str) -> dict:
+        """
+        Get request body from a request status file in s3.
+        :param request_id: Matrices concatenation request id.
+        :return: Request body.
+        """
+        key = request_id + JSON_SUFFIX
+
+        try:
+            content = s3_blob_store.get(bucket=REQUEST_STATUS_BUCKET_NAME, key=key)
+            body = json.loads(content)
+            return body
+        except Exception as e:
+            raise e
+
+    @staticmethod
+    def update_request(
+            bundle_uuids: List[str],
+            request_id: str,
+            job_id: str,
+            status: RequestStatus,
+            time_spent_to_complete="",
+            reason_to_abort="") -> None:
+        """
+        Update the request status json file in s3 bucket if exists. Otherwise,
+        create a new status file.
+
+        :param bundle_uuids: A list of bundle uuids.
+        :param request_id: Matrices concatenation request id.
+        :param job_id: Matrices concatenation job id.
+        :param status: Request status to update.
+        :param time_spent_to_complete: Time spent to complete the concatenation job.
+        :param reason_to_abort: Request abort reason.
+        """
+        request_key = request_id + JSON_SUFFIX
+
+        # Check whether request status file exists in the s3 bucket; Create one if not.
+        try:
+            RequestHandler.get_request_body(request_id=request_id)
+        except BlobNotFoundError:
+            with tempfile.TemporaryFile(dir=TEMP_DIR, suffix=JSON_SUFFIX) as f:
+                s3_blob_store.upload_file_handle(
+                    bucket=REQUEST_STATUS_BUCKET_NAME,
+                    key=request_key,
+                    src_file_handle=f
+                )
+        except Exception as e:
+            raise e
```

Comment:
Actually I think the effects of this are kind of severe. Making a GET for the object here means you don't get read-after-write consistency even on the first PUT.

Reply:
Yes, that's true. But even if I don't get read-after-write consistency on the first PUT, the worst thing that could happen is: 

 * Suppose two consecutive `update_request()` happens, and both of them having the same request id.  
 * The first `update_request()` GET the object from S3 and receives 404. Thus, it PUT a new status file into S3.
 * When the second `update_request()` tries GET the object from S3, (suppose the first PUT has not fully propagated) due to the eventual consistency, the second request stills receives 404. As a result, it will also PUT a new file into S3.

The main point of having a GET before PUT here is to avoid create redundant files in the local disk as there is a limitation on the disk size I can use for each lambda execution. Under the scenario I describe above, even though there are still redundant local files being created, it will not cause any bug, right? 

Or, do I miss something?

Reply:
My main concern here is that we're not getting anything out of this code block, and the function would work the same if we just deleted it. I think all this does is upload an empty file to S3, right? What's the benefit of that? If there aren't any benefits, then we don't need to investigate the costs too much.

Reply:
If I delete it, and directly create and upload a request status file to S3 each time when `update_request()` is called, then the following problem can happen:  

Suppose there is a request which asks for merging N matrices,  and when we download kth matrix, where k <= N, there is no more disk space available for that lambda invocation. As a result, the `OSError` will be caught in the following snippet of code:

```python
except Exception as e:
    # Update the request status to ABORT
     RequestHandler.update_request(
                matrix_format=self.suffix,
                bundle_uuids=bundle_uuids,
                request_id=request_id,
                job_id=job_id,
                status=RequestStatus.ABORT,
                reason_to_abort=traceback.format_exc()
      )

      raise e

However, since `update_request()` will attempt to create a new file, but there is actually no more space on the disk, the call to `update_request()` will cause another exception without setting the the request status to `ABORT`.

Reply:
I don't think I understand. I'm pretty sure `update_request` doesn't do any disk I/O except for the portion I'm proposing to delete.

Reply:
I am so sorry. It seems like I have a huge misunderstanding here. I originally think that I can only upload a file to S3 as a key. But I find that 

```python
boto3.resource("s3").Object()

will create one for me if it does not exist. I think you are totally right. I will delete that block of code.